python/cugraph/cugraph/tests/link_prediction/test_cosine.py

```diff
@@ -0,0 +1,368 @@
+EDGE_ATT_COL = "weight"
+MULTI_COL_SRC_0_COL = "src_0"
+MULTI_COL_DST_0_COL = "dst_0"
+MULTI_COL_SRC_1_COL = "src_1"
+MULTI_COL_DST_1_COL = "dst_1"
+
+
+# =============================================================================
+# Pytest Setup / Teardown - called for each test function
+# =============================================================================
+def setup_function():
+    gc.collect()
+
+
+# =============================================================================
+# Helper functions
+# =============================================================================
+def compare_cosine(cu_coeff, cpu_coeff, epsilon=1.0e-6):
+    assert len(cu_coeff) == len(cpu_coeff)
+    for i in range(len(cu_coeff)):
+        if np.isnan(cpu_coeff[i]):
+            assert np.isnan(cu_coeff[i])
+        elif np.isnan(cu_coeff[i]):
+            assert cpu_coeff[i] == cu_coeff[i]
+        else:
+            diff = abs(cpu_coeff[i] - cu_coeff[i])
+            # Properly handle floating-point arithmetic
+            assert diff <= abs(cu_coeff[i]) if \
+                (abs(cpu_coeff[i]) < abs(cu_coeff[i])) else \
+                    (abs(cpu_coeff[i]) * epsilon)
+
+def cugraph_call(benchmark_callable, graph_file, pairs, use_weight=False):
+    # Device data
+    G = graph_file.get_graph(
+        create_using=cugraph.Graph(directed=False), ignore_weights=not use_weight
+    )
+    # cugraph Cosine Call
+    df = benchmark_callable(cugraph.cosine, G, pairs, use_weight=use_weight)
+    df = df.sort_values(by=[VERTEX_PAIR_FIRST_COL, VERTEX_PAIR_SECOND_COL]).reset_index(
+        drop=True
+    )
+
+    return df[COSINE_COEFF_COL].to_numpy()
+
+
+
+def cosine(a, b, M):
+    # Retieve the out-degree of a
+    out_degree_a = M.indices[M.indptr[a]:M.indptr[a+1]]
+    # Retieve the out-degree of a
+    out_degree_b = M.indices[M.indptr[b]:M.indptr[b+1]]
+    
+    # Find the intersection of a and b
+    intersection_a_b = np.intersect1d(out_degree_a, out_degree_b)
+
+    norm_a = 0
+    norm_b = 0
+    a_dot_b = 0
+
+    for dst in intersection_a_b:
+        norm_a += pow(M[a, dst], 2)
+        norm_b += pow(M[b, dst], 2)
+        a_dot_b += (pow(M[a, dst], 2) * pow(M[b, dst], 2))
+
+    return a_dot_b / (pow(norm_a, 0.5) * pow(norm_b, 0.5)) if \
+        (norm_a * norm_b) != 0 else 0
+
+
+def cpu_call(M, first, second):
+    result = []
+    for i in range(len(first)):
+        result.append(cosine(first[i], second[i], M))
+    return result
+
+
+def compare(src1, dst1, val1, src2, dst2, val2):
+    #
+    #  We will do comparison computations by using dataframe
+    #  merge functions (essentially doing fast joins).  We
+    #  start by making two data frames
+    #
+    df1 = cudf.DataFrame()
+    df1["src1"] = src1
+    df1["dst1"] = dst1
+    if val1 is not None:
+        df1["val1"] = val1
+
+    df2 = cudf.DataFrame()
+    df2["src2"] = src2
+    df2["dst2"] = dst2
+    if val2 is not None:
+        df2["val2"] = val2
+
+    #
```

Comment:
Is there a reason why we can't just compare the two DFs at this point using something like `assert_frame_equal`?

Reply:
The test calling this function compares the result of `all_pairs_cosine_similarity`(with top_k = 10)  against `cosine_similarity` where their result is stored respectively in `df1` and `df2`. The `compare` function just ensures that `df1` is a subset of `df2` through the merge operation `df1.merge(df2, left_on=["src1", "dst1"], right_on=["src2", "dst2"])`. The test succeeds if the merge is a success (when the length of the merge's result is equal to the len(df1), otherwise, df1 and df2 will be merge in a way that highlight the difference which will be printed in this statement `if len(df1) != len(join):` for debugging purposes.